proposals/collection-literals.md

```diff
@@ -151,70 +156,72 @@ _Open issue: Relying on a parameter named `capacity` seems brittle. Is there an
+An error is reported for a collection literal if there are multiple `[CollectionBuilder]` attributes on the collection type.
 
-* For example, if an `init void Construct(T[] values)` instance method were added to [`ImmutableArray<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1), then it would be possible for the compiler to emit the following:
+The *create method* must be a `static` method defined on the *builder type*.
+The *create method* must have a single parameter of a type `System.ReadOnlySpan<T0>` passed by value, and a return value of the *collection type* to which the attribute is applied.
 
-    ```c#
-    T[] __storage = /* initialized using predefined rules */
-    ImmutableArray<T> __result = new ImmutableArray<T>();
-    __result.Construct(__storage);
-    ```
+The *builder type* cannot be generic.
+The *builder method* can be a generic *method* and the *arity* of the builder method must match the *arity* of the *collection type*.
 
-    `ImmutableArray<T>` would then take that array directly and use it as its own backing storage.  This would be safe because the compiler (following the requirements around `init`) would ensure that no other location in the code would have access to this temporary array, and thus it would not be possible to mutate it behind the back of the `ImmutableArray<T>` instance.
+Overloads with distinct signatures on the builder type, and overloads on base types, are ignored.
 
-    The above also demonstrates that this approach can work with struct types which do not have a [*parameterless struct constructor*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/parameterless-struct-constructors.md).  In the above, the call to `new ImmutableArray<T>()` is equivalent to `default(ImmutableArray<T>)`, (producing an `ImmutableArray<T>` whose [`IsDefault`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1.isdefault) property is initially true.  However, the `Construct` method can then safely update this to the final non-default state without that intermediate state being visible.
+An error is reported for a collection literal if the `[CollectionBuilder]` attribute on the collection type does not represent an invocable method with the expected signature defined on the *builder type*.
 
-* This formalization is quite beneficial because the only existing mechanism to (safely) create an ImmutableArray with values without copying is both excessively verbose and produces unavoidable garbage:
+If the parameter is `scoped`, the compiler *may* allocate the storage for the `ReadOnlySpan<T0>` on the stack rather than the heap.
 
-    ```c#
-    var __builder = ImmutableArray.CreateBuilder<int>(initialCapacity: __len);
+For example, a possible *create method* for `ImmutableArray<T>`:
+```csharp
+[CollectionBuilder(typeof(ImmutableArray), "Create")]
```

Comment:
This is striking me as `infoof`. Do we _really_ need to be able to specify member names here, and not just have a well-recognized pattern that the type indicated must follow?

Reply:
@333fred i believe we need this.  But i'll let @stephentoub weigh in.  I think the core issue is that the BCL expects (or already has) differently named methods in a few places, and they'd want these attributes to point at them (versus adding a new duplicative method).  Because of that, we need a type+methodname pair to clearly identify things.  

Note: i believe this is going through Runtime review soon (or maybe already has).  So i'm ok deferring to what they want.  ON our side we just need something to go find the appropriate method :)

Reply:
@333fred, are you suggesting that the compiler would look at all members of the specified type for one with the right shape and use whatever it finds?

Reply:
> @333fred, are you suggesting that the compiler would look at all members of the specified type for one with the right shape (accepting a `ReadOnlySpan<T>`, returning a `CollectionType`, name doesn't matter) and use whatever it finds?

Yes, but I would expect the name to be part of the pattern it looks for.

> I think the core issue is that the BCL expects (or already has) differently named methods in a few places, and they'd want these attributes to point at them (versus adding a new duplicative method).

We've already done this multiple times for language features. For example, `List<T>` added the `Slice` method which just defers to `GetRange`, to work with the slicing feature. I also wonder, if we have some small list of canonical methods, can we not just codify them now? Is there a list of the existing method names that we would want to consider for the pattern?

Reply:
> Yes, but I would expect the name to be part of the pattern it looks for.

That'll work for the specific uses planned for the core libraries for .NET 8, which are all in immutable collections.  It won't work when we add ctors to types like `HashSet<T>` and expect those to be usable.  It won't work when we add additional patterns and want multiple named methods on the same type, e.g. a CollectionsMarshal.CreateList and a CollectionsMarshal.CreateQueue.  And it won't work for other collections that already have an appropriately-shaped factory method with a different name, forcing the developer to add another API for it that does the same thing.

I think we should continue having the name be specified.  It's more flexible and specific and it doesn't hurt anything.

> For example, List<T> added the Slice method which just defers to GetRange, to work with the slicing feature

And we didn't like doing so.  We only did because `List<T>` is such a prominent type and it was the least friction-free way to unblock that use.  I do not see us, for example, adding a static `List` class with a static `Create` method.

Reply:
Ctors seem like something we codify. The multiple methods also seems like something we can also codify, with a rule involving the type name. Every time we've looked at something like `methodof` or `infoof` in C#, we've arrived at all sorts of issues that need to be resolved. I'm not innately opposed to it, but as this as historically been a difficult problem for us to address, we might want to consider a more hardcoded list.